// appium-xcuitest-driver/build/lib/device-log/ios-device-log.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IOSDeviceLog = void 0;
const support_1 = require("appium/support");
const ios_log_1 = require("./ios-log");
const appium_ios_device_1 = require("appium-ios-device");
const log = support_1.logger.getLogger('IOSDeviceLog');
class IOSDeviceLog extends ios_log_1.IOSLog {
    constructor(opts) {
        super();
        this.udid = opts.udid;
        this.showLogs = !!opts.showLogs;
        this.service = null;
    }
    /**
     * @override
     */
    async startCapture() {
        if (this.service) {
            return;
        }
        this.service = await appium_ios_device_1.services.startSyslogService(this.udid);
        this.service.start(this.onLog.bind(this));
    }
    onLog(logLine) {
        this.broadcast(logLine);
        if (this.showLogs) {
            log.info(logLine);
        }
    }
    /**
     * @override
     */
    get isCapturing() {
        return !!this.service;
    }
    /**
     * @override
     */
    // XXX: superclass is async, but this is not
    // eslint-disable-next-line require-await
    async stopCapture() {
        if (!this.service) {
            return;
        }
        this.service.close();
        this.service = null;
    }
}
exports.IOSDeviceLog = IOSDeviceLog;
exports.default = IOSDeviceLog;
